lib/kitchen/driver/gce.rb

```diff
@@ -16,154 +17,503 @@
+        operation = connection.insert_instance(project, zone, create_instance_object(server_name))
 
-      rescue Fog::Errors::Error, Excon::Errors::Error => ex
-        raise ActionFailed, ex.message
+        info("Zone operation #{operation.name} created. Waiting for it to complete...")
+        wait_for_operation(operation)
+
+        server              = server_instance(server_name)
+        state[:server_name] = server_name
+        state[:hostname]    = ip_address_for(server)
+        state[:zone]        = zone
+
+        info("Server <#{server_name}> created.")
+
+        update_windows_password(server_name)
+
+        info("Waiting for server <#{server_name}> to be ready...")
+        wait_for_server
+
+        info("GCE instance <#{server_name}> created and ready.")
+      rescue => e
+        error("Error encountered during server creation: #{e.class}: #{e.message}")
+        destroy(state)
+        raise
       end
 
       def destroy(state)
-        return if state[:server_id].nil?
+        @state      = state
+        server_name = state[:server_name]
+        return if server_name.nil?
+
+        unless server_exist?(server_name)
+          info("GCE instance <#{server_name}> does not exist - assuming it has been already destroyed.")
+          return
+        end
+
+        info("Destroying GCE instance <#{server_name}>...")
+        wait_for_operation(connection.delete_instance(project, zone, server_name))
+        info("GCE instance <#{server_name}> destroyed.")
 
-        instance = connection.servers.get(state[:server_id])
-        instance.destroy unless instance.nil?
-        info("GCE instance <#{state[:server_id]}> destroyed.")
-        state.delete(:server_id)
+        state.delete(:server_name)
         state.delete(:hostname)
+        state.delete(:zone)
       end
 
-      private
+      def validate!
+        raise "Either zone or region must be specified" unless config[:zone] || config[:region]
+        raise "'any' is no longer a valid region" if config[:region] == "any"
+        raise "Zone #{config[:zone]} is not a valid zone" if config[:zone] && !valid_zone?
+        raise "Region #{config[:region]} is not a valid region" if config[:region] && !valid_region?
+        raise "Machine type #{config[:machine_type]} is not valid" unless valid_machine_type?
+        raise "Disk type #{config[:disk_type]} is not valid" unless valid_disk_type?
+        raise "Network #{config[:network]} is not valid" unless valid_network?
+        raise "Email address of GCE user is not set" if winrm_transport? && config[:email].nil?
+
+        warn("Both zone and region specified - region will be ignored.") if config[:zone] && config[:region]
+      end
 
       def connection
-        options = {
-          provider: 'google',
-          google_client_email: config[:google_client_email],
-          google_project: config[:google_project]
-        }
+        return @connection unless @connection.nil?
```

Comment:
Could I ask you to set a custom user-agent please? I believe you'll need to set this as `application_name` and `application_version` when creating the client [[see here](https://github.com/google/google-api-ruby-client/blob/e8481dd14dc2af0667866a319832dc9adc8d54bf/lib/google/apis/options.rb#L32)].  If you already have a standard for TK, let me know what it is, otherwise maybe we can use something like `kitchen-google/x.y.z`?